src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,182 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+// This file describes AST nodes for authorization logic
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_H_
+#define SRC_IR_AUTH_LOGIC_AST_H_
+
+#include <string>
+#include <vector>
+#include <variant>
+
+namespace raksha::ir::auth_logic {
+
+class AstPrincipal {
+  public:
+    AstPrincipal(std::string name) : name_(std::move(name)) {}
+  private:
+    std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum AstSign { negated, positive };
+
+class AstPredicate {
+  public:
+    AstPredicate(
+        std::string name,
+        std::vector<std::string> args,
```

Comment:
What are these `args`? Should they have a bit more AST structure added to them?

Reply:
This predicate is `pred_name(arg_1, ..., arg_n)`, and the args are the `arg_1, ..., arg_n` bit. I don't see how more structure could be usefully added.